### PR TITLE
Fix errors in sequence deserialization

### LINKED ISSUE #342: Deserialization fails when choice and element are in complex element
Serialization works, but deserialization not:

```rust
#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "PascalCase")]
struct Root {
    element: Element,
    #[serde(rename = "$value")]
    choice: Choice,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct Element(String);

#[derive(Debug, serde::Serialize, serde::Deserialize)]
enum Choice {
    ChoiceOne(String),
    ChoiceTwo(String),
}

fn main() {
    let root = Root {
        element: Element("element".to_string()),
        choice: Choice::ChoiceTwo("choice_two".to_string()),
    };

    let s = quick_xml::se::to_string(&root).unwrap();
    println!("{:?}", s);
    let r: Root = quick_xml::de::from_str(&s).unwrap();
    println!("{:?}", r);
}
```

Serialization:
```xml
<Root><Element>element</Element><ChoiceTwo>choice_two</ChoiceTwo></Root>
```

Deserialization error:
```
'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom("unknown variant `Element`, expected `ChoiceOne` or `ChoiceTwo`")'
```

### PR DESCRIPTION
**Pull request by Mingun**
Replaces https://github.com/Mingun/fast-xml/pull/12

Fixes #342

This PR adds a bunch of tests to ensure, that deserialization of sequences is correct in all situations.

- Structs with `$value` field can now deserialize their other fields from elements, not only from attributes, as it was before:
    ```xml
    <any-tag>
      <before/>
      <item/>
      <item/>
      <item/>
      <after/>
    </any-tag>
    ```
    can be deserialized into
    ```rust
    struct AnyName {
      before: (),
      item: Vec<()>,
      after: (),
    }
    ```

- Added a new cargo feature `overlapped-lists` which allows also interleave sequence
  tags with non-sequence tags in the XML. When use that feature the following XML
  can be deserialized to the struct shown above:
    ```xml
    <any-tag>
      <item/>
      <item/>
      <before/>
      <item/>
      <after/>
      <item/>
    </any-tag>
    ```

### CODE REVIEW

File: tests/serde-de.rs [hunk 1]
```diff
@@ -512,6 +512,2288 @@ mod tuple_struct {
...
+    use super::*;
+
+    /// Check that top-level sequences can be deserialized from the multi-root XML documents
+    mod top_level {
+        use super::*;
+        use pretty_assertions::assert_eq;
+
+        #[test]
+        fn simple() {
+            from_str::<[(); 3]>("<root/><root>42</root><root>answer</root>").unwrap();
```
Comment (dralley) [inline tests/serde-de.rs:525]: Why not make this consistent with the Vec test?

File: tests/serde-de.rs [hunk 2]
```diff
@@ -512,6 +512,2530 @@ mod tuple_struct {
...
```
Comment (dralley) [inline tests/serde-de.rs:525]: This will panic if it cannot be deserialized to an array of size 3 yes?  (I am not as familiar with serde, my code doesn't use it)

Just curious why it's not asserted like the Vec scenario, I'm assuming this is why.

File: Cargo.toml [hunk 3]
```diff
@@ -42,6 +42,46 @@ default = []
...
+## ```ignore
+## #[derive(Deserialize)]
+## #[serde(rename_all = "kebab-case")]
+## struct AnyName {
+##   item: Vec<()>,
+##   another_item: (),
+## }
+## ```
+##
+## When feature is not enabled (default), only first element will be assotiated
```
Comment (dralley) [inline Cargo.toml:67]:
```suggestion
## When this feature is not enabled (default), only the first element will be associated
```

File: Cargo.toml [hunk 4]
```diff
@@ -42,6 +42,46 @@ default = []
 ## [standard compliant]: https://www.w3.org/TR/xml11/#charencoding
 encoding = ["encoding_rs"]
 
+## This feature enables support of deserializing of lists which tags are overlapped
```
Comment (dralley) [inline Cargo.toml:45]:
```suggestion
## This feature enables support for deserializing lists where tags are overlapped
```

File: Cargo.toml [hunk 5]
```diff
@@ -42,6 +42,46 @@ default = []
...
+## with tags that does not correspond to the list.
```
Comment (dralley) [inline Cargo.toml:46]:
```suggestion
## with tags that do not correspond to the list.
```

File: Cargo.toml [hunk 6]
```diff
@@ -42,6 +42,46 @@ default = []
...
+##
+## When this feature is enabled, that XML:
```
Comment (dralley) [inline Cargo.toml:48]:
```suggestion
## When this feature is enabled, the XML:
```

File: Cargo.toml [hunk 7]
```diff
@@ -42,6 +42,46 @@ default = []
...
+## #[derive(Deserialize)]
+## #[serde(rename_all = "kebab-case")]
+## struct AnyName {
+##   item: Vec<()>,
+##   another_item: (),
+## }
+## ```
+##
+## When feature is not enabled (default), only first element will be assotiated
+## with a field, and deserializer will report an error when it encounter a second
```
Comment (dralley) [inline Cargo.toml:68]:
```suggestion
## with the field, and the deserializer will report an error when it encounter a second
```

File: Cargo.toml (hunk 6 above)
Comment (dralley) [inline Cargo.toml:48]:
```suggestion
## When this feature is enabled, that XML:
```

File: Cargo.toml [hunk 8]
```diff
@@ -42,6 +42,46 @@ default = []
...
+## `<item/>`.
+##
+## Note, that enabling this feature can lead to high and even unlimited memory
+## consumption, because deserializer should check all events up to the end of a
+## container tag (`</any-name>` in that example) to figure out that there are no
+## more items for a field. If `</any-name>` or EOF even not encountered, the
```
Comment (dralley) [inline Cargo.toml:74]:
```suggestion
## more items for a field. If `</any-name>` or even EOF is not encountered, the
```

File: Cargo.toml [hunk 9]
```diff
@@ -42,6 +42,46 @@ default = []
...
+## parsing will never end which can lead to DoS.
```
Comment (dralley) [inline Cargo.toml:75]:
```suggestion
## parsing will never end which can lead to a denial-of-service (DoS) scenario.
```

Comment (dralley) [reply Cargo.toml:75]: I suppose that anyone with control over the XML could accomplish this regardless by just endlessly stuffing more items into the list, but this would be more difficult to notice.

File: Cargo.toml [hunk 10]
```diff
@@ -42,6 +42,46 @@ default = []
...
+##
+## Having several lists and overlapped elements for them in XML could also lead
+## to quadratic parsing time, because deserialzier have to check list of events
+## as many times as count of sequence fields present.
```
Comment (dralley) [inline Cargo.toml:79]:
```
## Having several lists and overlapped elements for them in XML could also lead
## to quadratic parsing time, because the deserializer must check the list of events
## as many times as there are sequence fields present in the schema.
```

**Comment by dralley** [comment]
This is a long PR so it might take me a few days to get through it.  Handling the documentation first since it gives me a better picture of what the code ought to look like.

File: src/de/mod.rs [hunk 11]
```diff
@@ -277,6 +279,11 @@ where
     /// moved from this to [`Self::read`].
     #[cfg(feature = "overlapped-lists")]
     write: VecDeque<DeEvent<'de>>,
+    /// Maximum number of events that could be skipped. This field is used to prevent
```
Comment (dralley) [inline src/de/mod.rs:282]:
```suggestion
/// Maximum number of events that can be skipped when processing sequences that occur out-of-order.
```

File: tests/serde-de.rs (hunk 2 above)
Comment (dralley) [reply tests/serde-de.rs:525]: It's fine

File: Cargo.toml [hunk 12]
```diff
@@ -78,8 +78,14 @@ encoding = ["encoding_rs"]
 ## to quadratic parsing time, because deserialzier have to check list of events
 ## as many times as count of sequence fields present.
 ##
+## To reduce negative consequences always [limit] maximum number of events that
```
Comment (dralley) [inline Cargo.toml:81]:
```suggestion
## To reduce negative consequences, always [limit] the maximum number of events that
```

File: src/errors.rs [hunk 13]
```diff
@@ -159,6 +161,10 @@ pub mod serialize {
         ExpectedStart,
         /// Unsupported operation
         Unsupported(&'static str),
+        /// Too many events was skipped during deserializing a sequence, limit of
```
Comment (dralley) [inline src/errors.rs:164]:
```suggestion
        /// Too many events were skipped while deserializing a sequence, event limit
```

File: src/errors.rs [hunk 14]
```diff
@@ -159,6 +161,10 @@ pub mod serialize {
...
+        /// events exceeded. The limit provides as an argument
```
Comment (dralley) [inline src/errors.rs:165]:
```suggestion
        /// exceeded. The limit was provided as an argument
```

File: src/de/mod.rs [hunk 15]
```diff
@@ -445,24 +520,36 @@ where
...
+                        _ => self.skip_event(event)?,
                     }
                 }
             }
             _ => Ok(()),
         }
     }
 
+    #[cfg(feature = "overlapped-lists")]
+    #[inline]
```
Comment (dralley) [inline src/de/mod.rs:542]: This is short enough that, since it's only used in the function above, it could be defined as a closure there.

No strong feelings about it.

File: src/de/mod.rs [hunk 16]
```diff
@@ -387,6 +396,72 @@ where
...
 
+    /// Set the maximum number of events that could be skipped during deserialization
+    /// of sequences.
+    ///
+    /// If `<element>` contains more than specified nested elements, `#text` or
+    /// CDATA nodes, then [`DeError::TooManyEvents`] will be returned during
+    /// deserialization of sequence field (any type that uses [`deserialize_seq`]
+    /// for the deserialization, for example, `Vec<T>`).
+    ///
+    /// This method can be used to prevent a [DoS] attack and infinity memory
```
Comment (dralley) [inline src/de/mod.rs:407]:
```suggestion
    /// This method can be used to prevent a [DoS] attack and infinite memory
```

or "unbounded"

File: src/de/mod.rs [hunk 17]
```diff
@@ -387,6 +396,72 @@ where
...
+    /// consumption when parse a very large or even an infinity XML to a sequence
```
Comment (dralley) [inline src/de/mod.rs:408]:
```suggestion
    /// consumption when parsing a very large or even an infinite XML to a sequence
```

Could you explain "infinite xml"?  (to me, not in the text)

File: src/de/mod.rs [hunk 18]
```diff
@@ -387,6 +396,72 @@ where
...
+    /// field.
+    ///
+    /// It is strongly recommended to set limit to some value when you parse data
+    /// from untrusted source. You should choose a value that you typical XMLs
```
Comment (dralley) [inline src/de/mod.rs:412]: "untrusted sources" or "an untrusted source"

Comment (dralley) [reply src/de/mod.rs:412]: your typical XML

File: src/de/mod.rs [hunk 19]
```diff
@@ -387,6 +396,72 @@ where
...
+    /// The XML that we try to parse look like this:
+    /// ```xml
+    /// <any-name>
+    ///   <item/>
+    ///   <!-- Bufferisation starts at this point -->
+    ///   <another-item>
+    ///     <some-element>with text</some-element>
+    ///     <yet-another-element/>
+    ///   </another-item>
+    ///   <!-- Buffer will be emptied at this point; was buffered 7 events -->
```
Comment (dralley) [inline src/de/mod.rs:433]:
```suggestion
    ///   <!-- Buffer will be emptied at this point; 7 events will be drained -->
```

File: src/de/mod.rs [hunk 20]
```diff
@@ -387,6 +396,72 @@ where
...
+    /// struct List {
+    ///   item: Vec<()>,
+    /// }
+    /// ```
+    ///
+    /// The XML that we try to parse look like this:
+    /// ```xml
+    /// <any-name>
+    ///   <item/>
+    ///   <!-- Bufferisation starts at this point -->
```
Comment (dralley) [inline src/de/mod.rs:428]:
```suggestion
    ///   <!-- Bufferization starts at this point -->
```

File: src/de/mod.rs (hunk 17 above)
Comment (Mingun) [reply src/de/mod.rs:408]: I mean, that someone could try to parse an XML stream what will never end, for example, some live event log:
```xml
<EventList>
  <Event/>
  <Event/>
  <Event/>
  ...
```
But without deserializing such stream in a streaming way it anyway impossible to get full list regardless of this option. I'll remove that sentence

File: src/de/mod.rs (hunk 19 above)
Comment (Mingun) [reply src/de/mod.rs:433]: It seems that suggested change will change an attention point slightly. I try to emphasize, that buffer must have size enough to store at least 7 events (which reader may unexpect, because from his point of view it may seem that only one element need to be skipped and calculations for possible buffer size can be totally wrong if follow that logic. The real size should be at least 7 times bigger), and after that change is seems that sentence mostly gives information that you get exactly 7 events. Maybe I'm wrong and that change does not change the sence... please tell me your opinion.

**Comment by Mingun** [comment]
@dralley , I've address all your suggestions except one (I have a question about it). Could you finish your review this week? I have a tons of plans and unfortunately they all are a strictly coupled with each other.

**Comment by dralley** [comment]
@Mingun Yeah, I'll take another look tonight

File: src/de/mod.rs (hunk 19 above)
Comment (dralley) [reply src/de/mod.rs:433]: You could reword it to "7 events were buffered" in that case

File: src/de/seq.rs [hunk 21]
```diff
@@ -1,18 +1,74 @@
...
+use crate::reader::Decoder;
 use serde::de::{self, DeserializeSeed};
+#[cfg(not(feature = "encoding"))]
+use std::borrow::Cow;
 
+/// Check if tag `start` is included in the `fields` list. `decoder` is used to
+/// get a string representation of a tag.
+///
+/// Returns `true`, if `start` is not in the `fields` list and `false` otherwise.
+pub fn is_unknown(
```
Comment (dralley) [inline src/de/seq.rs:12]: Could this be named more descriptively e.g. `is_unknown_{field|tag|element}`?

File: src/de/map.rs [hunk 22]
```diff
@@ -135,7 +140,24 @@ enum ValueSource {
...
+/// ```
+///
+/// Name of this tag is stored in a [`Self::start`] property.
+///
+/// # Lifetimes
+///
+/// - `'de` lifetime represents a buffer, from which deserialized values can
+///   borrow their data. Depending on the underlying reader, there can be an
+///   internal buffer of deserializer (i.e. deserializer itself) or an input
+///   (in that case it is possible to approach zero-cost deserialization).
```
Comment (dralley) [inline src/de/map.rs:160]: Zero-cost, or zero-copy?  The latter is what is specifically being achieved here while the former is kind of a fuzzy statement.  I take it to mean that "deserialization is fast enough that the CPU can keep up with IO" in the same vein as e.g. `simdjson`

File: src/de/seq.rs [hunk 23]
```diff
@@ -1,18 +1,74 @@
...
+
+    #[cfg(feature = "encoding")]
+    let tag = decoder.decode(start.name());
+
+    Ok(fields.iter().all(|&field| field != tag.as_ref()))
+}
+
+/// A filter that determines, what tags should form a sequence.
+///
+/// There is a two variant of sequences:
```
Comment (dralley) [inline src/de/seq.rs:28]: /// There are two variants of sequences:

File: src/de/map.rs [hunk 24]
```diff
@@ -43,12 +44,14 @@ enum ValueSource {
     /// </any-tag>
     /// ```
     Text,
-    /// Next value should be deserialized from an element with an any name.
-    /// Corresponding tag name will always be associated with a field with name
-    /// [`INNER_VALUE`].
+    /// Next value should be deserialized from an element with an any name, except
+    /// elements which name matched one of the struct fields. Corresponding tag
```
Comment (dralley) [inline src/de/map.rs:48]:
```suggestion
    /// elements with a name matching one of the struct fields. Corresponding tag
```

File: src/de/seq.rs (hunk 21 above)
Comment (dralley) [reply src/de/seq.rs:12]: "unknown" also sounds a little strange in this context since we theoretically do known the names of the enum variants?

re:

> 
>     enum Choice { one, two, three }
>     struct Root {
>         #[serde(rename = "$value")]
>         item: [Choice; 3],
>         node: (),
>     }
> 
> and this struct should be able to deserialized from
> 
>     <root>
>         <one/>
>         <two/>
>         <three/>
>         <node/>
>     </root>

Maybe `tag_maps_to_field`?  (That name sounds a bit wrong as well, so if you prefer the former names I'm fine with that)

File: src/de/seq.rs [hunk 25]
```diff
@@ -1,18 +1,74 @@
...
+/// - sequence where each element represented by tags with the same name
+/// - sequence where each element can have a different tag
+///
+/// The first variant could represent a collection of structs, the second --
+/// a collection of enums.
```
Comment (dralley) [inline src/de/seq.rs:33]:
```suggestion
/// a collection of enum variants
```

File: src/de/seq.rs (hunk 23 above)
Comment (dralley) [reply src/de/seq.rs:28]: To deconflict with "enum variants" perhaps "types of sequences" would be better here.

File: src/de/seq.rs [hunk 26]
```diff
@@ -1,18 +1,74 @@
...
+///
+/// In the second case we don't know what name sequence element will accept,
+/// so we take an any element. But because in XML sequences a flattened into a
+/// maps, then we could take an elements that have their own dedicated fields
+/// in a struct. To prevent this we use an `Exclude` filter, that filters out
+/// any known names of a struct fields.
```
Comment (dralley) [inline src/de/seq.rs:39]:
```suggestion
/// tags with any known field names from the struct.
```

File: src/de/seq.rs [hunk 27]
```diff
@@ -22,8 +78,10 @@ pub struct SeqAccess<'de, 'a, R>
 where
     R: XmlRead<'de>,
 {
+    /// Deserializer used to deserialize sequence items
     de: &'a mut Deserializer<'de, R>,
-    names: Names,
+    /// Filter that determines is that tag is a part of this sequence?
```
Comment (dralley) [inline src/de/seq.rs:83]:
```suggestion
    /// Filter that determines whether a tag is a part of this sequence
```

File: Changelog.md [hunk 28]
```diff
@@ -15,6 +15,9 @@
 - [#9]: Deserialization erroneously was successful in some cases where error is expected.
   This broke deserialization of untagged enums which rely on error if variant cannot be parsed
 - [#387]: Allow to have an ordinary elements together with a `$value` field
+- [#387]: Internal deserializer state can be broken when deserialize a map with
```
Comment (dralley) [inline Changelog.md:23]:
```suggestion
- [#387]: Internal deserializer state can be broken when deserializing a map with
```

File: Changelog.md [hunk 29]
```diff
@@ -15,6 +15,9 @@
...
+  a sequence field (such as `Vec<T>`), where elements of this sequence contains
+  another sequence. This error affects only users with a `serialize` feature enabled
```
Comment (dralley) [inline Changelog.md:25]:
```suggestion
  another sequence. This error affects only users with the `serialize` feature enabled
```

File: src/de/map.rs [hunk 30]
```diff
@@ -173,6 +173,14 @@ where
     source: ValueSource,
     /// List of field names of the struct. It is empty for maps
     fields: &'static [&'static str],
+    /// If `true`, then deserialized struct have a field with a special name
```
Comment (dralley) [inline src/de/map.rs:179]:
```suggestion
    /// If `true`, then the deserialized struct has a field with a special name:
```

File: src/de/seq.rs (hunk 21 above)
Comment (Mingun) [reply src/de/seq.rs:12]: I chose `is_unknown` simple in contrary to `is_known` (i.e. if tag is listed in `fields` it is known for a map), but because each usage of `is_known()` would be `!is_known()` I inverted it. The method does not have any specific cognitive load, so just `not_in` should be enough.

File: src/de/map.rs (hunk 22 above)
Comment (Mingun) [reply src/de/map.rs:160]: I mean that when optimized any serde deserializer code should gone as if you manually locate positions in the string and take a slice. But `zero-copy` is a better anyway.

File: src/de/seq.rs (hunk 23 above)
Comment (Mingun) [reply src/de/seq.rs:28]: Agreed

File: src/de/seq.rs (hunk 26 above)
Comment (Mingun) [reply src/de/seq.rs:39]: Then

> /// In the second case we don't know what tag name should be expected as a
> /// sequence element, so we accept any element. Since the sequence are flattened
> /// into maps, we _skip_ elements which have dedicated fields in a struct by using an 
> /// `Exclude` filter that filters out elements with names matching field names from the struct.

This describes that fact, that struct
```rust
struct AnyName {
  field: (),
  list: Vec<()>,
}
```
in XML would be represented as
```xml
<any-tag>
  <field/>
  <list/>
  <list/>
<any-tag>
```
The JSON equivalent would be
```json
{
  "field": null,
  "list": null,
  "list": null
}
```
instead of correct
```json
{
  "field": null,
  "list": [null, null]
}
```
Because collection element could be an `enum` and enums in XML could be represented as
```xml
<variant-name .../>
```
and because the type of a sequence element is opacue to us at the `AnyName` deserialization level, we should consider all tags as a potential list items with one exception: if we are know with guarantee that a tag with that name is mapped to another field. We list all names of that tags in the `Exclude` variant and this list is a `fields` list of a struct.
In order to achieve correct deserialization we should collect all tags into a sequence

File: src/de/map.rs [hunk 31]
```diff
@@ -364,6 +373,82 @@ where
     /// Access to the map that created this deserializer. Gives access to the
     /// context, such as list of fields, that current map known about.
     map: &'m mut MapAccess<'de, 'a, R>,
+    /// Determines, should [`Deserializer::next_text_impl()`] expand the second
+    /// level of tags or not.
+    ///
+    /// If this field is `true`, we processes the following XML shape:
```
Comment (dralley) [inline src/de/map.rs:390]:
```suggestion
    /// If this field is `true`, we process the following XML shape:
```

File: src/de/seq.rs (hunk 26 above)
Comment (dralley) [reply src/de/seq.rs]: Makes sense.  I always thought the most natural XML representation was akin to the JSON one, e.g.

```xml
<any-tag>
  <field/>
  <list>
    <item/>
    <item/>
    <item/>
  </list>
<any-tag>
```

But I can see some rationale for not doing it that way, too.

Comment (Mingun) [reply src/de/seq.rs]: Yes, it would be nice for this to be the case, but I often see such XML. If you think about it, this is natural, given how easy it is to describe such a structure in XSD:
```xml
<xs:complexType>
  <xs:sequence>
    <xs:element name="field"/>
    <xs:element name="list" maxOccurs="unbounded"/>
  </xs:sequence>
</xs:complexType>
```

File: .github/workflows/rust.yml [hunk 32]
```diff
@@ -40,6 +40,10 @@ jobs:
       env:
         LLVM_PROFILE_FILE: coverage/serialize-escape-html-%p-%m.profraw
       run: cargo test --features serialize,escape-html
+    - name: Run tests (all features)
+      env:
+        LLVM_PROFILE_FILE: coverage/all-features-%p-%m.profraw
+      run: cargo test --all-features
```
Comment (dralley) [inline .github/workflows/rust.yml:46]: Looks like this should cover testing both with and without `overlapped-lists` which is a concern because of so much alternate code :+1:

File: src/de/mod.rs [hunk 33]
```diff
@@ -370,12 +416,69 @@ where
...
+    /// read and write buffers.
+    ///
+    /// After calling this method, [`Self::peek()`] and [`Self::next()`] starts
+    /// return events that was skipped previously by calling [`Self::skip()`],
+    /// and only when all that events will be consumed, the deserializer starts
+    /// to drain events from underlying reader.
+    ///
+    /// This method MUST be called if any number of [`Self::skip()`] was called
+    /// after [`Self::new()`] or `start_replay()` or you'll lost events.
+    #[cfg(feature = "overlapped-lists")]
```
Comment (dralley) [inline src/de/mod.rs:563]: Just curious about the use case, is this basically for situations where you're parsing messy handwritten XML / HTML?  Generally I would think machine-generated XML wouldn't have this problem.

File: Cargo.toml [hunk 34]
```diff
@@ -42,6 +42,52 @@ default = []
 ## [standard compliant]: https://www.w3.org/TR/xml11/#charencoding
 encoding = ["encoding_rs"]
 
+## This feature enables support for deserializing lists where tags are overlapped
+## with tags that do not correspond to the list.
```
Comment (dralley) [inline Cargo.toml:46]: Just curious about the use case, is this basically for situations where you're parsing messy handwritten XML / HTML? Generally I would think machine-generated XML wouldn't have this problem.

It might be worth providing that as an example of a situation where you might want to enable it if that is the case.

File: .github/workflows/rust.yml (hunk 32 above)
Comment (Mingun) [reply .github/workflows/rust.yml:46]: Yes, that is why I add this.

**Review by dralley — APPROVED** [review]

File: Cargo.toml (hunk 34 above)
Comment (Mingun) [reply Cargo.toml:46]: Actually, I cannot remember that would I worked with such XMLs, but all parsers, that I know, very tolerant to overlapped lists (even in `xs:sequence`, although XSD requires exact order of tags. I never seen using `xs:all` that allows arbitrary order). So I thought there should be an option to allow this behavior

Comment (k-bx) [reply Cargo.toml:46]: @dralley I'm trying to parse KML file which can have elements of types Folder, Document and Point interleaved in a single list. I'd like to ideally get them in the same order, so I tried to create an enum covering all three cases, but it didn't work. So this at least gives me an option to parse it as three lists without the preservation of order in-between the items.

Comment (k-bx) [reply Cargo.toml:46]: @Mingun it works, but it isn't really what's needed here. What is the desired way to handle things:

input: `<TagOne></TagOne><TagTwo></TagTwo><TagOne></TagOne>`

parsed as:

```
struct TagOne {};
struct TagTwo {};
enum Tag {
  TagOne(TagOne),
  TagTwo(TagTwo)
}

-- result should be
vec![Tag::TagOne(TagOne {}), Tag::TagTwo(TagTwo {}), Tag::TagOne(TagOne {})]
```

I didn't see a way to do something like this via the current lib ^

Do I need to create a separate issue to make something like this possible?

Comment (Mingun) [reply Cargo.toml:46]: Yes, this is an error, I filled #500 for it
